pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java

```diff
@@ -0,0 +1,594 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.core.common.datatable;
+
+import com.google.common.primitives.Ints;
+import com.google.common.primitives.Longs;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.pinot.common.response.ProcessingException;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.common.utils.DataTable;
+import org.apache.pinot.common.utils.StringUtil;
+import org.apache.pinot.core.common.ObjectSerDeUtils;
+import org.apache.pinot.core.query.request.context.ThreadTimer;
+import org.apache.pinot.spi.utils.ByteArray;
+import org.apache.pinot.spi.utils.BytesUtils;
+
+import static org.apache.pinot.common.utils.DataTable.MetadataKeys.EXECUTION_THREAD_CPU_TIME_NS;
+import static org.apache.pinot.core.common.datatable.DataTableUtils.*;
+
+
+/**
+ * Datatable V3 implementation.
+ * The layout of serialized V3 datatable looks like:
+ * 	+-----------------------------------------------+
+ * 	| 13 bytes of header:                           |
+ * 	| VERSION                                       |
+ * 	| NUM_ROWS                                      |
+ * 	| NUM_COLUMNS                                   |
+ * 	| EXCEPTIONS SECTION START OFFSET               |
+ * 	| EXCEPTIONS SECTION LENGTH                     |
+ * 	| DICTIONARY_MAP SECTION START OFFSET           |
```

Comment:
Why are we treating EXCEPTIONS differently in V3 v/s V2 ? Nothing changes here right. They continue to be part of the metadata and along with rest of the metadata key-values, it moves to the end of the byte stream

Reply:
This is the only difference between V2 and V3 protocol right?

 ```
// VERSION
  // NUM_ROWS
  // NUM_COLUMNS
  // DICTIONARY_MAP (START|SIZE)
  // METADATA (START|SIZE) -> removed in V3 and moves to trailer/footer/end
  // DATA_SCHEMA (START|SIZE)
  // FIXED_SIZE_DATA (START|SIZE)
  // VARIABLE_SIZE_DATA (START|SIZE)
```

Reply:
V2 construct key for exception using "Exception"+errCode, e.g. Exception404, Exception500, then put into metadatam in V2 it's OK since V2 metadata key is String,  but in V3 it's impossible -- all key in V3 must be defined in enum statically. We can not construct a key dynamically

Reply:
Got it. Although it is possible to squeeze in the exceptions in the metadata section by mapping multiple exception codes to same ordinal, looks like this approach of having separate section of exceptions is cleaner. Discussed offline with @mqliang and @mcvsubbu . We all agreed with this